### DISCUSSION TITLE
ERR_SSL_WRONG_VERSION_NUMBER when running nextjs in docker cointainer with traefik

### DISCUSSION THREAD

**Discussion by Tassil0** [vercel/next.js#61730] [Help]
### Summary

Nextjs app running in docker, responds with 500 on every request on every route. The app is using traefik for SSL/reverse proxy, other services running the same traefik config have no issues whatsoever. Next server log only shows `ERR_SSL_WRONG_VERSION_NUMBER`.

### Additional information

Next server log:

```js
Error: fetch failed
    at context.fetch (/app/apps/webapp/.next/standalone/node_modules/.pnpm/next@14.0.2_@babel+core@7.23.2_react-dom@18.2.0_react@18.2.0_sass@1.66.1/node_modules/next/dist/server/web/sandbox/context.js:272:38)
    at Object.middleware_middleware [as handler] (/app/apps/webapp/.next/standalone/apps/webapp/.next/server/src/middleware.js:2253:15)
    at /app/apps/webapp/.next/standalone/apps/webapp/.next/server/src/middleware.js:1426:23
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (/app/apps/webapp/.next/standalone/apps/webapp/.next/server/src/middleware.js:1238:24)
    at adapter (/app/apps/webapp/.next/standalone/apps/webapp/.next/server/src/middleware.js:1413:53)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: [Error: D07B0FDCAAAA0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:
  ] {
    library: 'SSL routines',
    reason: 'wrong version number',
    code: 'ERR_SSL_WRONG_VERSION_NUMBER'
  }
}
```

The next app is built with option `output: standalone;` and run using `node server.js` command.
We use Next.js v14.

- Docker compose file, that is used to run the app: https://pastebin.com/VZWRAqzX
- `traefik.yaml`: https://pastebin.com/McXyKzZR
- Dockerfile: https://pastebin.com/60gGrwAY
- note: app is built in CI pipeline, then the build output is copied into the image
- app is using pnpm with turborepo monorepo
- other containers that run on this VPS under traefik have no issues
- traefik has no error logs

**Comment by noxify** [comment]
Had the same problem and found your post.

In our case, we have called an internal api endpoint via `fetch` to check the current session of an user ( we're using lucia-auth ).
We used `request.nextUrl.origin` initially to get the base url.
This worked fine locally, because the origin is/was `localhost`.

While deploying the app into our OpenShift cluster, the `origin` was the internal pod address.
This produces the mentioned error.

We solved it by setting an environment variable with the fallback to the `origin`.

```ts
 const origin = process.env.APPLICATION_URL
      ? // eslint-disable-next-line no-restricted-properties
        `https://${process.env.APPLICATION_URL}`
      : request.nextUrl.origin
```

**Comment by mr-chandan** [comment]
any solution ??

**Comment by mr-chandan** [comment]
i was using request.nextUrl.origin in my middleware.ts file i replaced it to http://127.0.0.1:3000 and it worked fine and also using http://localhost:3000 would not work

**Comment by hanssl** [comment]
same issue.  no solution yet?

**Comment by fernandowski** [comment]
for everyone who is having this issue. I do not think it is a bug that needs to be fixed by vercel what is happening you are likely making an api call using an unsupported protocol in your middleware. 

you can do something like this to figure out the protocol and host 

```
const host = request.headers.get("host");
const isLocalHost = host?.startsWith("localhost") || host?.startsWith("127.0.0.1");
const protocol = isLocalHost ? "http" : "https";
```
